File: backend/payment-service/src/main/java/com/skillinfinity/payment/entity/Payment.java
package com.skillinfinity.payment.entity;

import com.skillinfinity.payment.enumeration.PaymentGateway;
import com.skillinfinity.payment.enumeration.PaymentStatus;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "payments")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Payment {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(name = "user_id", nullable = false)
    private UUID userId;

    @Column(name = "payment_number", nullable = false, unique = true, length = 30)
    private String paymentNumber;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 30)
    @Builder.Default
    private PaymentStatus status = PaymentStatus.PENDING;

    @Column(name = "amount", nullable = false, precision = 15, scale = 2)
    private BigDecimal amount;

    @Column(name = "currency", nullable = false, length = 10)
    @Builder.Default
    private String currency = "CREDITS";

    @Column(name = "credits", nullable = false, precision = 15, scale = 2)
    private BigDecimal credits;

    @Enumerated(EnumType.STRING)
    @Column(name = "gateway", length = 30)
    private PaymentGateway gateway;

    @Column(name = "gateway_payment_id", length = 100)
    private String gatewayPaymentId;

    @Column(name = "gateway_order_id", length = 100)
    private String gatewayOrderId;

    @Column(name = "gateway_signature", length = 255)
    private String gatewaySignature;

    @Column(name = "description", length = 500)
    private String description;

    @Column(name = "reference_id", length = 100)
    private String referenceId;

    @Column(name = "reference_type", length = 50)
    private String referenceType;

    @Column(name = "coupon_code", length = 50)
    private String couponCode;

    @Column(name = "discount_amount", precision = 15, scale = 2)
    @Builder.Default
    private BigDecimal discountAmount = BigDecimal.ZERO;

    @Column(name = "tax_amount", precision = 15, scale = 2)
    @Builder.Default
    private BigDecimal taxAmount = BigDecimal.ZERO;

    @Column(name = "total_amount", precision = 15, scale = 2)
    private BigDecimal totalAmount;

    @Column(name = "failure_reason", length = 500)
    private String failureReason;

    @Column(name = "failure_code", length = 50)
    private String failureCode;

    @Column(name = "refunded_amount", precision = 15, scale = 2)
    @Builder.Default
    private BigDecimal refundedAmount = BigDecimal.ZERO;

    @Column(name = "paid_at")
    private LocalDateTime paidAt;

    @Column(name = "refunded_at")
    private LocalDateTime refundedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "created_by", length = 36)
    private String createdBy;

    @Column(name = "updated_by", length = 36)
    private String updatedBy;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        updatedAt = LocalDateTime.now();
        if (totalAmount == null) {
            totalAmount = amount;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/src/types/payment.ts
/**
 * Payment domain types — mirror the payment-service DTOs
 * (PaymentResponse, InvoiceResponse, ReceiptResponse, TransactionResponse,
 * PaymentRequest, PaymentConfirmationRequest, CouponRequest, RefundRequest).
 */

export type PaymentStatus =
  | 'PENDING'
  | 'PROCESSING'
  | 'COMPLETED'
  | 'FAILED'
  | 'REFUNDED'
  | 'CANCELLED';

export type PaymentGateway =
  | 'INTERNAL'
  | 'STRIPE'
  | 'RAZORPAY'
  | 'PAYPAL'
  | 'UPI'
  | 'CARD';

export interface Payment {
  id: string;
  userId: string;
  paymentNumber?: string;
  status: PaymentStatus | string;
  amount: number;
  currency?: string;
  credits?: number;
  discountAmount?: number;
  taxAmount?: number;
  totalAmount?: number;
  description?: string;
  referenceId?: string;
  referenceType?: string;
  couponCode?: string;
  gateway?: string;
  gatewayPaymentId?: string;
  gatewayOrderId?: string;
  failureReason?: string;
  refundedAmount?: number;
  paidAt?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface Invoice {
  id: string;
  invoiceNumber?: string;
  paymentId: string;
  userId?: string;
  status?: string;
  amount: number;
  discountAmount?: number;
  taxAmount?: number;
  totalAmount?: number;
  currency?: string;
  description?: string;
  issuedAt?: string;
  dueDate?: string;
  paidAt?: string;
}

export interface Receipt {
  id: string;
  receiptNumber?: string;
  paymentId?: string;
  invoiceId?: string;
  userId?: string;
  amount: number;
  currency?: string;
  paymentMethod?: string;
  gatewayTransactionId?: string;
  issuedAt?: string;
}

export interface PaymentTransaction {
  id: string;
  transactionNumber?: string;
  transactionType?: string;
  status?: string;
  amount: number;
  currency?: string;
  description?: string;
  referenceId?: string;
  createdAt?: string;
}

/* ---------------- Request payloads (mirror backend) ---------------- */

export interface PaymentInitRequest {
  amount: number;
  /** Number of credits purchased with this payment (pack credits, not INR). */
  credits?: number;
  currency?: string;
  description?: string;
  referenceId?: string;
  referenceType?: string;
  couponCode?: string;
  gateway?: string;
  subscriptionPlanId?: string;
}

export interface PaymentConfirmationRequest {
  paymentId: string;
  gatewayPaymentId?: string;
  gatewayOrderId?: string;
  gatewaySignature?: string;
  gatewayTransactionId?: string;
}

export interface PaymentFailureRequest {
  paymentId: string;
  failureReason?: string;
  failureCode?: string;
  gatewayResponse?: string;
}

export interface RefundRequest {
  paymentId: string;
  amount: number;
  reason: string;
}

export interface CouponValidationRequest {
  couponCode: string;
  amount: number;
  userId?: string;
}

/* ---------------- Subscriptions ---------------- */

export interface SubscriptionPlan {
  id: string;
  name: string;
  /** Audience this plan targets — 'LEARNER' or 'MENTOR'. */
  type?: 'LEARNER' | 'MENTOR';
  description?: string;
  price: number;
  currency?: string;
  durationDays?: number;
  maxSessionsPerMonth?: number;
  features?: string[];
  active?: boolean;
}

export interface MySubscription {
  subscriptionId: string;
  plan: SubscriptionPlan;
  status: string;
  startedAt?: string;
  expiresAt?: string;
  autoRenew?: boolean;
}

export interface SubscriptionPurchaseRequest {
  planId: string;
  couponCode?: string;
}

/* ---------------- Razorpay (INR, test mode) ---------------- */

/** Backend-controlled credit package — the backend decides credits + price. */
export interface CreditPackage {
  code: string;
  name: string;
  credits: number;
  price: number;
  currency?: string;
  highlighted?: boolean;
  features?: string[];
  active?: boolean;
}

/** Razorpay order details for opening the checkout. */
export interface RazorpayOrder {
  paymentId: string;
  orderId: string;
  /** Amount in paise (₹109 → 10900). */
  amount: number;
  currency: string;
  /** Razorpay Key ID — safe for the browser. */
  keyId: string;
  credits?: number;
  discountAmount?: number;
  totalAmount?: number;
}

export interface RazorpayVerifyRequest {
  paymentId: string;
  razorpayPaymentId?: string;
  razorpayOrderId?: string;
  razorpaySignature: string;
}

/** Razorpay subscription checkout details (order-based, same as credit purchase). */
export interface RazorpaySubscriptionCheckout {
  paymentId: string;
  orderId: string;
  keyId: string;
  planId?: string;
  planName?: string;
  amount: number;
  currency: string;
  totalCount?: number;
}

export interface RazorpaySubscriptionVerifyRequest {
  paymentId: string;
  razorpayPaymentId?: string;
  razorpayOrderId?: string;
  razorpaySignature: string;
}

export interface CreditPurchaseRequest {
  packageCode: string;
  couponCode?: string;
}
